File: android-uxsdk-beta-visualcamera/src/main/java/dji/ux/beta/visualcamera/widget/cameraconfig/iso/CameraConfigISOAndEIWidgetModel.java
/*
 * Copyright (c) 2018-2020 DJI
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package dji.ux.beta.visualcamera.widget.cameraconfig.iso;

import androidx.annotation.NonNull;

import dji.common.camera.ExposureSettings;
import dji.common.camera.SettingsDefinitions;
import dji.keysdk.CameraKey;
import dji.keysdk.DJIKey;
import dji.thirdparty.io.reactivex.Flowable;
import dji.thirdparty.io.reactivex.functions.Function;
import dji.thirdparty.org.reactivestreams.Publisher;
import dji.ux.beta.core.base.DJISDKModel;
import dji.ux.beta.core.base.WidgetModel;
import dji.ux.beta.core.communication.ObservableInMemoryKeyedStore;
import dji.ux.beta.core.util.DataProcessor;
import dji.ux.beta.core.util.SettingDefinitions;

/**
 * Widget Model for the {@link CameraConfigISOAndEIWidget} used to define
 * the underlying logic and communication
 */
public class CameraConfigISOAndEIWidgetModel extends WidgetModel {

    //region Constants
    /**
     * The value to display when the ISO is locked.
     */
    protected static final String LOCKED_ISO_VALUE = "500";
    //endregion

    //region Fields
    private final DataProcessor<ExposureSettings> exposureSettingsProcessor;
    private final DataProcessor<SettingsDefinitions.ISO> isoProcessor;
    private final DataProcessor<SettingsDefinitions.ExposureSensitivityMode> exposureSensitivityModeProcessor;
    private final DataProcessor<Integer> eiValueProcessor;
    private final DataProcessor<String> isoAndEIValueProcessor;
    private int cameraIndex;
    private SettingsDefinitions.LensType lensType = SettingsDefinitions.LensType.ZOOM;
    //endregion

    //region Constructor
    public CameraConfigISOAndEIWidgetModel(@NonNull DJISDKModel djiSdkModel,
                                           @NonNull ObservableInMemoryKeyedStore keyedStore) {
        super(djiSdkModel, keyedStore);
        this.cameraIndex = SettingDefinitions.CameraIndex.CAMERA_INDEX_0.getIndex();
        exposureSettingsProcessor = DataProcessor.create(new ExposureSettings(SettingsDefinitions.Aperture.UNKNOWN,
                SettingsDefinitions.ShutterSpeed.UNKNOWN,
                0,
                SettingsDefinitions.ExposureCompensation.UNKNOWN));
        isoProcessor = DataProcessor.create(SettingsDefinitions.ISO.UNKNOWN);
        exposureSensitivityModeProcessor = DataProcessor.create(SettingsDefinitions.ExposureSensitivityMode.UNKNOWN);
        eiValueProcessor = DataProcessor.create(0);
        isoAndEIValueProcessor = DataProcessor.create("");
    }
    //endregion

    //region Data

    /**
     * Get the current index of the camera the widget model is reacting to
     *
     * @return current camera index
     */
    @NonNull
    public SettingDefinitions.CameraIndex getCameraIndex() {
        return SettingDefinitions.CameraIndex.find(cameraIndex);
    }

    /**
     * Set the index of the camera for which the widget model should react
     *
     * @param cameraIndex camera index
     */
    public void setCameraIndex(@NonNull SettingDefinitions.CameraIndex cameraIndex) {
        this.cameraIndex = cameraIndex.getIndex();
        restart();
    }

    /**
     * Get the current type of the lens the widget model is reacting to
     *
     * @return current lens type
     */
    @NonNull
    public SettingsDefinitions.LensType getLensType() {
        return lensType;
    }

    /**
     * Set the type of the lens for which the widget model should react
     *
     * @param lensType lens type
     */
    public void setLensType(@NonNull SettingsDefinitions.LensType lensType) {
        this.lensType = lensType;
        restart();
    }

    /**
     * Get the ISO.
     *
     * @return Flowable for the DataProcessor that user should subscribe to.
     */
    public Flowable<SettingsDefinitions.ISO> getISO() {
        return isoProcessor.toFlowable();
    }

    /**
     * Get either the ISO or the exposure index value as a displayable String.
     *
     * @return Flowable for the DataProcessor that user should subscribe to.
     */
    public Flowable<String> getISOAndEIValue() {
        return isoAndEIValueProcessor.toFlowable();
    }

    /**
     * Get whether the camera is in EI mode.
     *
     * @return Flowable for the DataProcessor that user should subscribe to.
     */
    public Flowable<Boolean> isEIMode() {
        return exposureSensitivityModeProcessor.toFlowable()
                .concatMap((Function<SettingsDefinitions.ExposureSensitivityMode, Publisher<Boolean>>) exposureSensitivityMode ->
                        Flowable.just(exposureSensitivityMode == SettingsDefinitions.ExposureSensitivityMode.EI));
    }
    //endregion

    //region LifeCycle
    @Override
    protected void inSetup() {
        DJIKey exposureSettingsKey = djiSdkModel.createLensKey(CameraKey.EXPOSURE_SETTINGS, cameraIndex, lensType.value());
        DJIKey isoKey = djiSdkModel.createLensKey(CameraKey.ISO, cameraIndex, lensType.value());
        DJIKey exposureSensitivityModeKey = CameraKey.create(CameraKey.EXPOSURE_SENSITIVITY_MODE, cameraIndex);
        DJIKey eiValueKey = CameraKey.create(CameraKey.EI_VALUE, cameraIndex);

        bindDataProcessor(exposureSettingsKey, exposureSettingsProcessor);
        bindDataProcessor(isoKey, isoProcessor);
        bindDataProcessor(exposureSensitivityModeKey, exposureSensitivityModeProcessor);
        bindDataProcessor(eiValueKey, eiValueProcessor);
    }

    @Override
    protected void inCleanup() {
        //Nothing to cleanup
    }

    @Override
    protected void updateStates() {
        updateConsolidatedISOValue();
    }
    //endregion

    //region Helpers
    private void updateConsolidatedISOValue() {
        if (exposureSensitivityModeProcessor.getValue() == SettingsDefinitions.ExposureSensitivityMode.EI) {
            isoAndEIValueProcessor.onNext(String.valueOf(eiValueProcessor.getValue()));
        } else {
            if (isoProcessor.getValue() == SettingsDefinitions.ISO.FIXED && exposureSettingsProcessor.getValue().getISO() == 0) {
                isoAndEIValueProcessor.onNext(LOCKED_ISO_VALUE);
            } else {
                isoAndEIValueProcessor.onNext(String.valueOf(exposureSettingsProcessor.getValue().getISO()));
            }
        }
    }
    //endregion
}

File: android-uxsdk-beta-visualcamera/src/main/java/dji/ux/beta/visualcamera/widget/cameraconfig/iso/CameraConfigISOAndEIWidgetModel.kt
/*
 * Copyright (c) 2018-2021 DJI
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */
package dji.ux.beta.visualcamera.widget.cameraconfig.iso

import dji.common.camera.ExposureSettings
import dji.common.camera.SettingsDefinitions
import dji.common.camera.SettingsDefinitions.ExposureSensitivityMode
import dji.common.camera.SettingsDefinitions.LensType
import dji.keysdk.CameraKey
import dji.keysdk.DJIKey
import dji.thirdparty.io.reactivex.Flowable
import dji.thirdparty.io.reactivex.functions.Consumer
import dji.ux.beta.core.base.DJISDKModel
import dji.ux.beta.core.base.SchedulerProvider
import dji.ux.beta.core.base.WidgetModel
import dji.ux.beta.core.communication.ObservableInMemoryKeyedStore
import dji.ux.beta.core.module.LensModule
import dji.ux.beta.core.util.DataProcessor
import dji.ux.beta.core.util.SettingDefinitions.CameraIndex

/**
 * The value to display when the ISO is locked.
 */
private const val LOCKED_ISO_VALUE = "500"
private const val TAG = "ConfigISOEIWidMod"

/**
 * Widget Model for the [CameraConfigISOAndEIWidget] used to define
 * the underlying logic and communication
 */
class CameraConfigISOAndEIWidgetModel(
        djiSdkModel: DJISDKModel,
        keyedStore: ObservableInMemoryKeyedStore
) : WidgetModel(djiSdkModel, keyedStore) {
    //region Fields
    private val cameraConnectionDataProcessor: DataProcessor<Boolean> = DataProcessor.create(false)
    private val isoEIStateProcessor: DataProcessor<CameraConfigISOEIState> = DataProcessor.create(CameraConfigISOEIState.ProductDisconnected)
    private val exposureSettingsProcessor: DataProcessor<ExposureSettings> = DataProcessor.create(ExposureSettings(SettingsDefinitions.Aperture.UNKNOWN,
            SettingsDefinitions.ShutterSpeed.UNKNOWN,
            0,
            SettingsDefinitions.ExposureCompensation.UNKNOWN))
    private val isoProcessor: DataProcessor<SettingsDefinitions.ISO> = DataProcessor.create(SettingsDefinitions.ISO.UNKNOWN)
    private val exposureSensitivityModeProcessor: DataProcessor<ExposureSensitivityMode> = DataProcessor.create(ExposureSensitivityMode.UNKNOWN)
    private val eiValueProcessor: DataProcessor<Int> = DataProcessor.create(0)
    private val lensModule = LensModule()

    /**
     * Index of the camera the widget model is reacting to.
     */
    var cameraIndex: CameraIndex = CameraIndex.find(CameraIndex.CAMERA_INDEX_0.index)
        set(value) {
            field = value
            lensModule.setCameraIndex(this, value)
            restart()
        }

    /**
     * Type of the lens the widget model is reacting to.
     */
    var lensType: LensType = LensType.ZOOM
        set(value) {
            field = value
            restart()
        }
    //endregion
    //region Data

    /**
     * ISO/EI state
     */
    val isoEIState: Flowable<CameraConfigISOEIState>
        get() = isoEIStateProcessor.toFlowable()


    //endregion
    //region LifeCycle
    init {
        addModule(lensModule)
    }

    override fun inSetup() {
        val cameraConnectionKey: DJIKey = CameraKey.create(CameraKey.CONNECTION, cameraIndex.index)
        val exposureSettingsKey: DJIKey = lensModule.createLensKey(CameraKey.EXPOSURE_SETTINGS, cameraIndex.index, lensType.value())
        val isoKey: DJIKey = lensModule.createLensKey(CameraKey.ISO, cameraIndex.index, lensType.value())
        val exposureSensitivityModeKey: DJIKey = CameraKey.create(CameraKey.EXPOSURE_SENSITIVITY_MODE, cameraIndex.index)
        val eiValueKey: DJIKey = CameraKey.create(CameraKey.EI_VALUE, cameraIndex.index)
        bindDataProcessor(cameraConnectionKey, cameraConnectionDataProcessor)
        bindDataProcessor(exposureSettingsKey, exposureSettingsProcessor)
        bindDataProcessor(isoKey, isoProcessor)
        bindDataProcessor(exposureSensitivityModeKey, exposureSensitivityModeProcessor)
        bindDataProcessor(eiValueKey, eiValueProcessor)
        addDisposable(lensModule.isLensArrangementUpdated()
                .observeOn(SchedulerProvider.io())
                .subscribe(Consumer { value: Boolean ->
                    if (value) {
                        restart()
                    }
                }, logErrorConsumer(TAG, "on lens arrangement updated")))
    }

    override fun inCleanup() {
        //Nothing to cleanup
    }

    override fun updateStates() {
        if (productConnectionProcessor.value) {
            if(cameraConnectionDataProcessor.value) {
                if (exposureSensitivityModeProcessor.value == ExposureSensitivityMode.EI) {
                    isoEIStateProcessor.onNext(CameraConfigISOEIState.EI(eiValueProcessor.value.toString()))
                } else {
                    when {
                        isoProcessor.value == SettingsDefinitions.ISO.FIXED && exposureSettingsProcessor.value.iso == 0 -> {
                            isoEIStateProcessor.onNext(CameraConfigISOEIState.FIXED(LOCKED_ISO_VALUE))
                        }
                        isoProcessor.value == SettingsDefinitions.ISO.AUTO -> {
                            isoEIStateProcessor.onNext(CameraConfigISOEIState.AUTO(exposureSettingsProcessor.value.iso.toString()))
                        }
                        else -> {
                            isoEIStateProcessor.onNext(CameraConfigISOEIState.ISO(exposureSettingsProcessor.value.iso.toString()))
                        }
                    }
                }
            } else {
                isoEIStateProcessor.onNext(CameraConfigISOEIState.CameraDisconnected)
            }
        } else {
            isoEIStateProcessor.onNext(CameraConfigISOEIState.ProductDisconnected)
        }
    }

    //endregion

    /**
     * Class to represent states of ISO/EI
     */
    sealed class CameraConfigISOEIState {
        /**
         *  When product is disconnected
         */
        object ProductDisconnected : CameraConfigISOEIState()

        /**
         *  When camera is disconnected
         */
        object CameraDisconnected : CameraConfigISOEIState()

        /**
         * Exposure sensitivity mode is EI
         */
        data class EI(val eiValue: String) : CameraConfigISOEIState()

        /**
         * Exposure sensitivity mode is AUTO ISO
         */
        data class AUTO(val isoValue: String) : CameraConfigISOEIState()

        /**
         * Exposure sensitivity mode is FIXED ISO
         */
        data class FIXED(val isoValue: String) : CameraConfigISOEIState()

        /**
         * Exposure sensitivity mode is ISO
         */
        data class ISO(val isoValue: String) : CameraConfigISOEIState()
    }
}